### PR TITLE
Refactor functorialisation

### LINKED ISSUE #41: Add iterated functorialisation
There is no obstruction to functorialise multiple times with respect to one or several variables. This should be added to the syntax. For instance, it is expected that one could write any of the following terms
- `comp[[a]] x`
- `comp[[a]] [f]`
- `comp [[a]] [[b]]`

### PR DESCRIPTION
**Pull request by thibautbenjamin**
- Refactor the data structure for functorialisation
- Simplify the api for the functorialisation module
- Allow for performing iterated functorialisations

fix #41

### CODE REVIEW

**Comment by thibautbenjamin** [comment]
@regular-citizen : I have done the refactoring of the functorialisation. It was easier than expected, and it let me implement the iterated functorialisation very easily. It's ready for review now :)

File: lib/functorialisation.ml [hunk 1]
```diff
@@ -3,6 +3,26 @@ open Kernel.Unchecked_types
 
 exception NonMaximal of string
 exception FunctorialiseMeta
+exception WrongNumberOfArguments
+
+let list_functorialised l c =
+  let rec list l c =
```
Comment (regular-citizen) [inline lib/functorialisation.ml:9]: This is impressively cryptic code.

File: lib/functorialisation.ml [hunk 2]
```diff
@@ -20,39 +40,52 @@ let ctx_one_var c x =
...
-    let ctx_base = Unchecked.ps_to_ctx ps in
-    let ctx,assocs = ctx ctx_base l in
-    let tm = Coh (Cohdecl(ps,ty),(Unchecked.identity_ps ctx_base)) in
-    let tm_f = Unchecked.tm_apply_sub tm (target_subst assocs) in
-    let ty = Arr (ty, tm, tm_f) in
-    let ps = Kernel.PS.(forget (mk (Kernel.Ctx.check ctx))) in
-    let _, names,_ = Unchecked.db_levels ctx in
-    let ty = Unchecked.rename_ty ty names in
-    Cohdecl(ps, ty),ps
+let coh_one_step ps ty l =
```
Comment (regular-citizen) [inline lib/functorialisation.ml:54]: I really don't understand what this is doing

**Comment by thibautbenjamin** [comment]
Ok well, i will push some comments to indicate what all these are. It's pretty simple code overall, but kind of requires to have the data structures we are using in mind. I gotbtoo used on working on this on my own and forgot that I should actually document my code

**Comment by thibautbenjamin** [comment]
@regular-citizen I pushed a new commit with a bunch of comments. It's a bit hard to explain what everything is doing while staying concise, but hopefully it should be more readable now. Don't hesitate to ask if anything is still unclear!

**Pull request merged**